views/default/elgg/Ajax.js

```diff
@@ -106,38 +129,77 @@ define(function (require) {
 				}
 			}
 
-			if ($.isArray(options.success)) {
-				throw new Error('The array form of options.success is not supported');
-			}
-
-			if (elgg.isFunction(options.success)) {
-				options.success = function (data) {
-					data = unwrap_data(data);
-					orig_options.success(data);
-				};
-			}
-
 			if (use_spinner) {
 				options.beforeSend = function () {
-					orig_options.beforeSend && orig_options.beforeSend();
+					orig_options.beforeSend && orig_options.beforeSend.apply(null, arguments);
 					spinner.start();
 				};
 				options.complete = function () {
 					spinner.stop();
-					orig_options.complete && orig_options.complete();
+					orig_options.complete && orig_options.complete.apply(null, arguments);
 				};
 			}
 
 			if (!options.error) {
-				options.error = elgg.ajax.handleAjaxError;
+				options.error = function (jqXHR, textStatus, errorThrown) {
+					if (!jqXHR.getAllResponseHeaders()) {
+						// user aborts (like refresh or navigate) do not have headers
+						return;
+					}
+
+					try {
+						var data = $.parseJSON(jqXHR.responseText);
+						if ($.isPlainObject(data)) {
+							extract_metadata(data);
+						}
+					} catch (e) {
+						if (window.console) {
+							console.warn(e.message);
+						}
+					}
+
+					if (!error_displayed) {
+						elgg.register_error(elgg.echo('ajax:error'));
+					}
+				};
 			}
 
+			options.dataFilter = function (data, type) {
+				if (type !== 'json') {
+					return data;
+				}
+
+				data = $.parseJSON(data);
+
+				extract_metadata(data);
+
+				var params = {
+					options: orig_options
+				};
+				if (hook_type) {
+					data = elgg.trigger_hook(Ajax.RESPONSE_DATA_HOOK, hook_type, params, data);
+				}
+
+				jqXHR.ajaxData = data;
```

Comment:
Why not add `jqXHR.errorDisplayed` and consider it a fix for https://github.com/Elgg/Elgg/issues/9767.  We just need a temporary solution to determine if error was shown.